src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs

```diff
@@ -33,84 +31,70 @@ internal static class KeyAnalyzer
-                    {
-                        results = CreateAnalysisResults(
-                            uniqueStrings, ignoreCase, minLength, maxLength, index, count,
-                            static (string s, int index, int count) => s.AsSpan(index, count));
-                        return true;
-                    }
-                }
-
-                // There were no left-justified substrings of this length available.
-                // If all of the strings are of the same length, then just checking left-justification is sufficient.
-                // But if any strings are of different lengths, then we'll get different alignments for left- vs
-                // right-justified substrings, and so we also check right-justification.
-                if (minLength != maxLength)
-                {
-                    // toggle the direction and re-use the comparer and hashset (HasSufficientUniquenessFactor clears it)
-                    comparer.IsLeft = false;
-
-                    // For each index, get a uniqueness factor for the right-justified substrings.
+                    // For each index, get a uniqueness factor for the left-justified substrings.
                     // If any is above our threshold, we're done.
                     for (int index = 0; index <= minLength - count; index++)
                     {
-                        // Get a uniqueness factor for the right-justified substrings.
-                        // If it's above our threshold, we're done.
-                        comparer.Index = -index - count;
+                        comparer.Index = index;
+
                         if (HasSufficientUniquenessFactor(set, uniqueStrings))
                         {
-                            results = CreateAnalysisResults(
-                                uniqueStrings, ignoreCase, minLength, maxLength, comparer.Index, count,
-                                static (string s, int index, int count) => s.AsSpan(s.Length + index, count));
-                            return true;
+                            return CreateAnalysisResults(
+                                uniqueStrings, ignoreCase, minLength, maxLength, index, count);
+                        }
+                    }
+
+                    // There were no left-justified substrings of this length available.
+                    // If all of the strings are of the same length, then just checking left-justification is sufficient.
+                    // But if any strings are of different lengths, then we'll get different alignments for left- vs
+                    // right-justified substrings, and so we also check right-justification.
+                    if (minLength != maxLength)
+                    {
+                        // For each index, get a uniqueness factor for the right-justified substrings.
+                        // If any is above our threshold, we're done.
+                        comparer.Index = -count;
```

Comment:
Here's the real logical improvement. We do the math for setting a negative index (e.g. from the right side) starting with `count` characters from the right (as before). Then on [line  78](https://github.com/dotnet/runtime/pull/88516/files#diff-8dfef3756fbf6ca0aab7668e6abf54ae6c416c60cb1b3bd86455204942f7dbefR78) we keep decrementing `comparer.Index` as we go. which is used by the `Slicer`